### ISSUE TITLE
Campaign read and write over API

### ISSUE THREAD

**Issue by pastorhudson** [mikeckennedy/listmonk#8]
I currently use wagtail for my content management. When I write a blog post I save a template to mail chimp so that I can easily send this as an email. I'm looking at switching to listmonk and want to have the same feature. It looks fairly straightforward to implement Would basic template read and write be welcome here?

**Comment by mikeckennedy** [comment]
Hey. Sounds interesting. Can you give me a sense of the steps the library needs to perform? Basically, adding the ability to create or edit a template? or is that a campaign in listmonk terms?

**Comment by pastorhudson** [comment]
I think you're right. I haven't used listmonk yet. I wanted to get this working before I killed my mailchimp.
Here's what I found.

- Templates are html + CSS with placeholders for subject, content, unsubscribe etc.
- Campaigns are created and a template is selected. The content is then the body of the campaign object.
- I think I want to be able to create / sync campaigns and add the body of my blog post to the campaign when I click publish in my cms. If I make edits in my cms I want it to update the campaign content as well.

This is basically how I do it in mailchimp.

```
import mailchimp_marketing as MailchimpMarketing
from django.template.loader import render_to_string
from mailchimp_marketing.api_client import ApiClientError
from django.conf import settings


def get_mailchimp_templates(template_name):
    """

    Retrieve the unique identifier (id) for a Mailchimp template by template name.

    Parameters:
    template_name (str): The name of the Mailchimp template to search for.

    Returns:
    str or None: The unique identifier (id) of the Mailchimp template if found, otherwise None.
    """
    try:
        client = MailchimpMarketing.Client()
        client.set_config({
            "api_key": settings.MAILCHIMP_API_KEY,
            "server": settings.MAILCHIMP_DATA_CENTER
        })

        response = client.templates.list()
        for template in response['templates']:
            if template['name'] == template_name:
                return template['id']
        return None
    except ApiClientError as error:
        print("Error: {}".format(error.text))


def update_template(template_id, template_name, page):
    """
    Update a template with a new name and HTML content.

    Parameters:
    template_id (str): The ID of the template to update.
    template_name (str): The new name for the template.
    page (dict): The data to be rendered in the HTML template.

    Returns:
    None
    """
    try:
        client = MailchimpMarketing.Client()
        client.set_config({
            "api_key": settings.MAILCHIMP_API_KEY,
            "server": settings.MAILCHIMP_DATA_CENTER
        })
        html_body = render_to_string("chefblog/email/email_blog.html", {"page": page})

        response = client.templates.update_template(template_id, {"name": template_name, "html": html_body})
    except ApiClientError as error:
        print("Error: {}".format(error.text))


def send_to_mailchimp(page):
    """

    Send data to Mailchimp API.

    Parameters:
    - page: (object) Page object containing data to be sent to Mailchimp.

    Returns:
    None

    """
    try:
        client = MailchimpMarketing.Client()
        client.set_config({
            "api_key": settings.MAILCHIMP_API_KEY,
            "server": settings.MAILCHIMP_DATA_CENTER
        })

        template_id = get_mailchimp_templates(page.title)
        if template_id:
            update_template(template_id, page.title, page)
        else:
            html_body = render_to_string("chefblog/email/email_blog.html", {"page": page})
            response = client.templates.create({"name": page.title, "html": html_body})
    except ApiClientError as error:
        print("Error: {}".format(error.text))


```

I think I would want to be able to create a campaign with the content of my cms blog post.
We need to get campaigns (Read)
Create Campaigns (Write)
Edit Campaigns

The sync method could be managed by the library user. I could do this with requests pretty easily, but it might make for a nice addition to this library. I'm willing to work on a PR if you like. Also I'm a huge fan of the podcast :)

**Comment by pastorhudson** [comment]
Motivation: I always forget to send email. I always say I'll get around to it, but I don't. Usually the email is a blog post. So when I started just automatically adding templates to mailchimp then I could go create campaigns with just a few clicks. It maid a big difference for me. I don't even care that I won't send all blog posts as email. Just having it in my mail software is a win.

**Comment by mikeckennedy** [comment]
> I haven't used listmonk yet. I wanted to get this working before I killed my mailchimp.

Good plan. :)

> Also I'm a huge fan of the podcast :)

Thanks a bunch!

> We need to get campaigns (Read), Create Campaigns (Write), Edit Campaigns

Yes, that sounds correct to me. The term template in listmonk is more of a design wrapper like top level nav, footer, etc. Campaign = single newsletter/email using that template.

I think this would be nice. I'd love a PR if you're up for it.

**Comment by pastorhudson** [comment]
Thanks so much for letting me take a shot at contributing! I have listmonk running and I have basic create, read, delete, and preview working for campaigns. I just need to add update and make sure everything is clean.

I have each feature as separate commits on my fork currently. 

Do you prefer I squash my commits to one commit before submitting PR? 

Should I submit PR on main branch? Or something else?

**Comment by pastorhudson** [comment]
Update is complete. I think the PR is ready for submission.

**Comment by mikeckennedy** [comment]
Great, just send the PR to main and I'll give it a look. Do you like the way it's working compared to mailchimp?

**Comment by pastorhudson** [comment]
I have integrated my subscribers, and campaign auto creation. It works fantastic I was able to send all my current blog posts in as campaigns. I kinda wish I could search campaigns by tag but that's a listmonk issue. As far as just listmonk vs mailchimp. I much prefer listmonk. Mailchimp is charging me monthly and I don't even send mass email monthly. . . Listmonk is dead simple and I like it that way.

**Comment by mikeckennedy** [comment]
Thanks again @pastorhudson This is all good sans the typing issue I mentioned in PR #9 Once you get that sorted (not sure which way it should go - dicts or ints), I'll push all these changes to PyPI under v0.3.0.

**Comment by pastorhudson** [comment]
I think I got it sorted in PR #12  Hopefully it makes sense.

**Comment by mikeckennedy** [comment]
Published this to pypi as version 0.3.0. Thanks for all the effort @pastorhudson
